rp2040-hal/src/adc.rs

```diff
@@ -1,3 +1,104 @@
 //! Analog-Digital Converter (ADC)
 // See [Chapter 4 Section 9](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
 // TODO
+
+use hal::adc::{Channel, OneShot};
+use pac::{ADC, RESETS};
+
+use crate::{
+    gpio::Pin,
+    gpio::{
+        bank0::{Gpio26, Gpio27, Gpio28, Gpio29},
+        FloatingInput,
+    },
+    resets::SubsystemReset,
+};
+
+/// Adc
+pub struct Adc {
+    device: ADC,
+}
+
+impl Adc {
+    /// Create new adc struct and bring up adc
+    pub fn new(device: ADC, resets: &mut RESETS) -> Self {
+        device.reset_bring_down(resets);
+        device.reset_bring_up(resets);
+
+        // Enable adc and temp sensor
+        device.cs.write(|w| w.en().set_bit());
+
+        // Wait for adc ready
+        while !device.cs.read().ready().bit_is_set() {
+            cortex_m::asm::nop();
```

Comment:
Performing a nop in a spin loop does not accomplish anything.
Unless you're going to put the processor to sleep waiting for interrupt (wfi) you should leave the body of the while empty.